src/app/modules/form/public-form/public-form.middlewares.ts

```diff
@@ -1,9 +1,52 @@
 import { RequestHandler } from 'express'
 import { StatusCodes } from 'http-status-codes'
+import mongoose from 'mongoose'
 
+import { createLoggerWithLabel } from '../../../../config/logger'
 import { WithForm } from '../../../../types'
+import getSubmissionModel from '../../../models/submission.server.model'
 import { FormDeletedError } from '../form.errors'
-import { isFormPublic } from '../form.service'
+import { deactivateForm, isFormPublic } from '../form.service'
+
+const SubmissionModel = getSubmissionModel(mongoose)
+
+const logger = createLoggerWithLabel(module)
+
+/**
+ * Express middleware function that checks if a form has exceeded its submission limits before allowing
+ * downstream middleware to handle the request. Otherwise, it returns a HTTP 404.
+ */
+export const checkFormSubmissionLimitAndDeactivate: RequestHandler = (
+  req,
+  res,
+  next,
+) => {
+  const { form } = req as WithForm<typeof req>
+  if (form.hasSubmissionLimit) {
+    return SubmissionModel.countDocuments({ form: form._id })
+      .exec()
+      .then((queryResults) => {
+        if (queryResults >= form.submissionLimit) {
+          logger.warn({
+            message:
+              'Form submission reached maximum submission count, deactivating form.',
+            meta: { form: form._id },
+          })
+          return deactivateForm(form._id).then(() => {
+            return res.status(StatusCodes.NOT_FOUND).json({
+              message: form.inactiveMessage,
+              isPageFound: true, // Flag to prevent default 404 subtext ("please check link") from showing
+              formTitle: form.title,
+            })
+          })
+        } else {
+          return next()
+        }
```

Comment:
> src/app/modules/form/public-form/public-form.middlewares.ts(28,13): error TS2345: Argument of type '(queryResults: number) => void | Promise<Response<any, Record<string, any>, number>>' is not assignable to parameter of type '(value: number) => Response<any, Record<string, any>, number> | PromiseLike<Response<any, Record<string, any>, number>>'.

The build error is complaining that the function `(queryResults) => {...}` is not of the correct type because TypeScript performs type inference when it sees `return deactivateForm(...)` and `return res.status(...)` and they don't match what is expected in the first `then` block.

Try moving the return statements to after the function calls complete? Or even better, port this to the modern async/await syntax.